cdap-watchdog/src/main/java/io/cdap/cdap/logging/appender/CompositeLogAppender.java

```diff
@@ -0,0 +1,120 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.logging.appender;
+
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.Context;
+import com.google.common.annotations.VisibleForTesting;
+import io.cdap.cdap.logging.framework.local.LocalLogAppender;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A log appender that delegates logging events to a list of other log appenders.
+ */
+public class CompositeLogAppender extends LogAppender {
+
+  private static final org.slf4j.Logger LOG = LoggerFactory.getLogger(CompositeLogAppender.class);
+  private final List<LogAppender> appenders;
+
+  /**
+   * Constructs a CompositeLogAppender with the given list of appenders.
+   *
+   * @param appenders the list of {@link LogAppender} instances to delegate logging events to
+   */
+  public CompositeLogAppender(List<LogAppender> appenders) {
+    if (appenders == null) {
+      throw new IllegalArgumentException("Appenders list cannot be null");
+    }
+
+    // Make the appenders list unmodifiable to ensure thread safety and prevent external
+    // modifications.
+    this.appenders = Collections.unmodifiableList(new ArrayList<>(appenders));
+    setName(getClass().getName());
+  }
+
+  @Override
+  public void start() {
+    appenders.forEach(appender -> safelyExecute(
+        appender::start,
+        String.format("Failed to start appender: %s", appender.getName())
+    ));
+    super.start();
+  }
+
+  @Override
+  public void stop() {
+    super.stop();
+    appenders.forEach(appender -> safelyExecute(
+        appender::stop,
+        String.format("Failed to stop appender: %s", appender.getName())
+    ));
+  }
+
+  @Override
+  protected void appendEvent(LogMessage logMessage) {
+    appenders.forEach(appender -> safelyExecute(
+        () -> appender.appendEvent(logMessage),
+        String.format("Failed to append log message to appender: %s. Message: %s",
+            appender.getName(), logMessage.getFormattedMessage())
+    ));
+  }
+
+  @Override
+  public void setContext(Context context) {
+    super.setContext(context);
+    appenders.forEach(appender -> safelyExecute(
+        () -> appender.setContext(context),
+        String.format("Failed to set context to appender: %s", appender.getName())
+    ));
+  }
+
+  @Override
+  public synchronized void doAppend(ILoggingEvent eventObject) {
+    if (shouldSkipLogging()) {
```

Comment:
Instead of executing `shouldSkipLogging()` each time `doAppend()` is called, can the result be cached?

Reply:
Can the result of `shouldSkipLogging()` be stored in a variable and reused?

If there a better solution to remove this complexity altogether?